### PR TITLE
Type settings

### PR DESCRIPTION
**Pull request by Fripe070**
### Checklist
- [ ] I have planned and discussed this feature with other contributors.
- [x] I have run basic tests on my code to ensure it is functional.
- [ ] I have thoroughly tested my code with various states, contexts and edge cases.
- [x] I have searched for a similar pull request in this repository and didn't find anything relevant.

### Summary

Cleans up a lot of typing issues, and makes `breadcord.config.Setting` into a generic.
This means that doing something like this would now be marked as invalid by the type checker due to a type mismatch.
```py
my_setting = cast(breadcord.config.Setting[bool], self.settings.my_setting)
@my_setting.observe
def observer(_, new: str) -> None:
    ...
```

### CODE REVIEW

File: breadcord/config.py [hunk 1]
```diff
@@ -73,10 +74,12 @@ def root(self) -> SettingsGroup:
         node = self
         while node.parent is not None:
             node = node.parent
-        return node
+        if __debug__ and not isinstance(node, SettingsGroup):
```
Comment (Fripe070) [inline breadcord/config.py:78]: This feels a bit questionable, and I'd like some feedback on what we think of it

Comment (m31-galaxy) [reply breadcord/config.py:78]: Because I'm braindead right now, what does this achieve?

File: breadcord/config.py [hunk 2]
```diff
@@ -145,7 +150,7 @@ def observe(
         This method can be used as a decorator, with optional parentheses for arguments.
 
         :param observer: The callback function. Takes two parameters ``old`` and ``new``, which correspond to the value
-            of the setting before and after it is updateed respectively.
+            of the setting before and after it is updated respectively.
```
Comment (m31-galaxy) [inline breadcord/config.py:154]: oops

**Review by m31-galaxy — DISMISSED** [review]

File: breadcord/config.py (hunk 1 above)
Comment (Fripe070) [reply breadcord/config.py:78]: It makes the type checker happy by eliminating the case where the root is not a settings group. This should never happen and I’ve hence only made it run in debug

Comment (m31-galaxy) [reply breadcord/config.py:78]: Might be worth adding a comment stating that it's an impossible branch then, otherwise would cause confusion when reading ("how could I end up in a state where this isn't SettingsGroup?")

Comment (Fripe070) [reply breadcord/config.py:78]: Maybe (I say, knowing full and well I will forget about this in 5 minutes)

**Review by alf1e — APPROVED** [review]
LGTM
